### PR TITLE
Fetch go version using gimme if needed

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Allows kube::golang::verify_go_version to choose a go version. Defaults to .go-version, can be overridden by GO_VERSION=1.x or skipped entirely with FORCE_HOST_GO=y

Copies approach taken by https://github.com/kubernetes-sigs/kind/blob/main/hack/build/setup-go.sh

xref https://github.com/kubernetes/test-infra/issues/28310

#### Does this PR introduce a user-facing change?
```release-note
The go version defined in `.go-version` is now fetched when invoking test, build, and code generation targets if the current go version does not match it. Set $FORCE_HOST_GO=y while testing or building to skip this behavior, or set $GO_VERSION to override the selected go version.
```

/cc @dims @BenTheElder

### CODE REVIEW

**Comment by dims** [comment]
definitely makes things easier. we should probably not get in the way of the local dev env as well (only warn them?) also not install something unless they explicitly ok us installing something in their env?

File: hack/lib/golang.sh [hunk 1]
```diff
@@ -497,6 +497,15 @@ EOF
 #   env-var GO15VENDOREXPERIMENT=1
 #   current directory is within GOPATH
 kube::golang::setup_env() {
+  # default GO_VERSION to content of .go-version
+  GO_VERSION="${GO_VERSION:-"$(cat "${KUBE_ROOT}/.go-version")"}"
```
Comment (aojea) [inline hack/lib/golang.sh:501]: should this be a local variable?

Comment (aojea) [reply hack/lib/golang.sh:501]: ah, ok, you allow to override it

Comment (liggitt) [reply hack/lib/golang.sh:501]: yeah, probably, will clean this up when we're happy with the overall shape

**Comment by liggitt** [comment]
> also not install something unless they explicitly ok us installing something in their env?

that's my instinct as well... could make our CI jobs opt into this behavior, and indicate how to opt-into auto-version-setup in verify_go_version

**Comment by liggitt** [comment]
awesome, this caught an issue in typecheck running with go1.20rc3 that https://github.com/kubernetes/kubernetes/pull/114502 missed:

> ERROR(linux/ppc64le): /home/prow/go/src/k8s.io/kubernetes/_output/local/.gimme/versions/go1.20rc3.linux.amd64/src/os/user/cgo_lookup_unix.go:169:22: invalid argument: index startSize (variable of type bufferKind) must be integer

we already exclude some cgo things in the typecheck because the go type scanner can't reliably interpret it... added a commit to exclude error checking of stdlib packages, will open in a separate PR to merge ahead of https://github.com/kubernetes/kubernetes/pull/114502

**Comment by BenTheElder** [comment]
I think we should consider opt-out.

If you want a custom go toolchain you already need to override any of the dockerized bits.

opt-out gives a smoother more reproducible developer experience for the 99% case.

KIND has opt-out + it will just use the existing toolchain if the version is correct.

**Comment by BenTheElder** [comment]
I think the tricky part in k/k is aligning these knobs with the dockerized build bits.
IIRC kind builds are fully using dockerized but some other CI may be mixed.

**Comment by liggitt** [comment]
interesting... by changing nothing other than the .go-version, this was able to build with go1.20rc3 even in the e2e:

> I0129 05:08:21.171] Server Version: version.Info{Major:"1", Minor:"27+", GitVersion:"v1.27.0-alpha.1.79+87116b4ea03d66", GitCommit:"87116b4ea03d66a1723a16c6571931edf73d8b5f", GitTreeState:"clean", BuildDate:"2023-01-28T18:20:43Z", GoVersion:"go1.20rc3", Compiler:"gc", Platform:"linux/amd64"}

that's ... kind of awesome

File: hack/lib/golang.sh [hunk 2]
```diff
@@ -491,12 +491,27 @@ EOF
...
+  # only setup go if we haven't set FORCE_HOST_GO, or `go version` doesn't match GO_VERSION
+  if ! ([ -n "${FORCE_HOST_GO:-}" ] || \
+      (command -v go >/dev/null && [ "$(go version | cut -d' ' -f3)" = "go${GO_VERSION}" ])); then
+      export GIMME_ENV_PREFIX=${GIMME_ENV_PREFIX:-"${KUBE_OUTPUT}/.gimme/envs"}
+      export GIMME_VERSION_PREFIX=${GIMME_VERSION_PREFIX:-"${KUBE_OUTPUT}/.gimme/versions"}
+      # eval because the output of this is shell to set PATH etc.
```
Comment (BenTheElder) [inline hack/lib/golang.sh:511]: i think you'll have to add a shellcheck exception here

Comment (liggitt) [reply hack/lib/golang.sh:511]: `verify: shellcheck` passed... maybe it shouldn't have?

**Comment by liggitt** [comment]
kind run:
> Server Version: version.Info{Major:"1", Minor:"27+", GitVersion:"v1.27.0-alpha.1.166+6436f1ca94e3f5", GitCommit:"6436f1ca94e3f55d917e98b644b579ea2539e9f4", GitTreeState:"clean", BuildDate:"2023-01-31T20:06:58Z", GoVersion:"go1.20rc3", Compiler:"gc", Platform:"linux/amd64"}

🎉

File: hack/lib/golang.sh (hunk 2 above)
Comment (liggitt) [reply hack/lib/golang.sh:511]: oh, this file already had `# shellcheck disable=SC2034 # Variables sourced in other scripts.` at the top

**Comment by liggitt** [comment]
/hold
for reviews and an opt-in vs opt-out decision

**Comment by liggitt** [comment]
some thoughts on opt-in vs opt-out:

* we already fetch builder images for a specified go version, fetching a specific go version seems similar in character
* the gimme-fetched version gets stored under `${KUBE_OUTPUT}` by default, so it does not pollute the user's environment
* as Ben noted, propagating opt-in options into our containerized build was... painful
* defaulting to obtaining the go version the repo says it uses makes a lot of the non-containerized scripts (update-gofmt, update-vendor, etc) work better for developers by default

**Comment by liggitt** [comment]
all presubmits passed on go1.20, logs verified builds and runs actually used go1.20 🎉

**Comment by dims** [comment]
> all presubmits passed on go1.20, logs verified builds and runs actually used go1.20 🎉

Sweet!!

**Comment by liggitt** [comment]
dropped .go-version commit, ready for review

File: hack/lib/golang.sh [hunk 3]
```diff
@@ -464,7 +464,21 @@ kube::golang::create_gopath_tree() {
...
 # Ensure the go tool exists and is a viable version.
+# Inputs:
+#   env-var GO_VERSION is the desired go version to use, downloading it if needed (defaults to content of .go-version)
+#   env-var FORCE_HOST_GO set to a non-empty value uses the go version in the $PATH and skips ensuring $GO_VERSION is used
 kube::golang::verify_go_version() {
+  # default GO_VERSION to content of .go-version
+  GO_VERSION="${GO_VERSION:-"$(cat "${KUBE_ROOT}/.go-version")"}"
+  # only setup go if we haven't set FORCE_HOST_GO, or `go version` doesn't match GO_VERSION
+  if ! ([ -n "${FORCE_HOST_GO:-}" ] || \
+      (command -v go >/dev/null && [ "$(go version | cut -d' ' -f3)" = "go${GO_VERSION}" ])); then
```
Comment (aojea) [inline hack/lib/golang.sh:475]: the code to detect the go version in line 491 appends GOFLAGS='', is that something that affects the output?

**Comment by aojea** [comment]
LGTM
/assign @dims @BenTheElder @thockin

**Comment by dims** [comment]
/priority important-soon
/triage accepted
/approve
/lgtm

thanks for adding the `FORCE_HOST_GO` flag!

**Comment by BenTheElder** [comment]
belated
/lgtm
/approve
very happy to see this!

**Pull request merged**